lib/include/pl/core/ast/ast_node_type_decl.hpp

```diff
@@ -76,15 +76,46 @@ namespace pl::core::ast {
         [[nodiscard]] std::vector<std::shared_ptr<ptrn::Pattern>> createPatterns(Evaluator *evaluator) const override {
             evaluator->updateRuntime(this);
 
-            evaluator->pushTemplateParameters();
-            ON_SCOPE_EXIT { evaluator->popTemplateParameters(); };
+            std::vector<std::unique_ptr<ASTNodeLiteral>> templateParamLiterals(this->m_templateParameters.size());
 
-            for (const auto &templateParameter : this->m_templateParameters) {
+            // Get template parameter values before pushing a new template parameter scope so that variables from the parent scope are used before being overwritten.
+            for (size_t i = 0; i < this->m_templateParameters.size(); i++) {
+                auto &templateParameter = this->m_templateParameters[i];
                 if (auto lvalue = dynamic_cast<ASTNodeLValueAssignment *>(templateParameter.get())) {
+                    if (!lvalue->getRValue())
+                        err::E0003.throwError(fmt::format("There is no value for template parameter variable {}. This is a bug.", lvalue->getLValueName()), {}, this);
                     auto value = lvalue->getRValue()->evaluate(evaluator);
-                    if (auto literal = dynamic_cast<ASTNodeLiteral*>(value.get()); literal != nullptr) {
-                        evaluator->createVariable(lvalue->getLValueName(), new ASTNodeBuiltinType(literal->getValue().getType()), {}, false, false, true);
-                        evaluator->setVariable(lvalue->getLValueName(), literal->getValue());
+                    if (auto literal = dynamic_cast<ASTNodeLiteral*>(value.release()))
+                        templateParamLiterals[i] = std::unique_ptr<ASTNodeLiteral>(literal);
+                    else
+                        err::E0003.throwError(fmt::format("Template parameter {} is not a literal. This is a bug.", lvalue->getLValueName()), {}, this);
+                }
+            }
+
+            evaluator->pushTemplateParameters();
+            ON_SCOPE_EXIT {
+                evaluator->popTemplateParameters();
+            };
+
+            {
+                evaluator->pushSectionId(ptrn::Pattern::HeapSectionId);
+                ON_SCOPE_EXIT {
+                    evaluator->popSectionId();
+                };
+
+                for (size_t i = 0; i < this->m_templateParameters.size(); i++) {
```

Comment:
Same here

Reply:
This uses an index to look up the values in `m_templateParameters` and `templateParameterLiterals` at the same time, should I change this to use a for-each and an iterator declared outside the loop?